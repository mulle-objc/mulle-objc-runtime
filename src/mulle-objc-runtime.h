//
//  mulle_objc.h
//  mulle-objc-runtime
//
//  Created by Nat! on 10.07.16.
//  Copyright (c) 2016 Nat! - Mulle kybernetiK.
//  Copyright (c) 2016 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_objc_runtime_h__
#define mulle_objc_runtime_h__

/*
 * If you want to include <mulle-objc-runtime/mulle-objc-runtime.h> you
 * should be compiling with an Objective-C compiler, that defines the
 * following defines properly. Do not use .c  use .m with mulle-objc-runtime.h
 */

// catch this early
#if ! defined( __MULLE_OBJC_TPS__) && ! defined( __MULLE_OBJC_NO_TPS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_TPS__ or __MULLE_OBJC_NO_TPS__)"
#endif
#if defined( __MULLE_OBJC_TPS__) && defined( __MULLE_OBJC_NO_TPS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_TPS__ and __MULLE_OBJC_NO_TPS__)"
#endif
#if ! defined( __MULLE_OBJC_FCS__) && ! defined( __MULLE_OBJC_NO_FCS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_FCS__ or __MULLE_OBJC_NO_FCS__)"
#endif
#if defined( __MULLE_OBJC_FCS__) && defined( __MULLE_OBJC_NO_FCS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_FCS__ and __MULLE_OBJC_NO_FCS__)"
#endif
#if ! defined( __MULLE_OBJC_TAO__) && ! defined( __MULLE_OBJC_NO_TAO__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_TAO__ or __MULLE_OBJC_NO_TAO__)"
#endif
#if defined( __MULLE_OBJC_TAO__) && defined( __MULLE_OBJC_NO_TAO__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_TAO__ and __MULLE_OBJC_NO_TAO__)"
#endif

//
// these should have been defined by the compiler already, but for
// mulle-objc-runtime testing with gcc its convenient to add these if missing.
//
#ifndef __MULLE_OBJC_UNIVERSEID__
# define __MULLE_OBJC_UNIVERSEID__  0
#endif
#ifndef __MULLE_OBJC_UNIVERSENAME__
# define __MULLE_OBJC_UNIVERSENAME__  NULL
#endif

#if defined( __MULLE_OBJC_TPS__) && __MULLE_OBJC_UNIVERSEID__ != 0
# error "TPS can only be used in the default universe not in " #__MULLE_OBJC_UNIVERSENAME__
#endif


#include "include.h"

#include "mulle-metaabi.h"

#include "mulle-objc-atomicpointer.h"
#include "mulle-objc-builtin.h"
#include "mulle-objc-call.h"
#include "mulle-objc-class.h"
#include "mulle-objc-classpair.h"
#include "mulle-objc-class-convenience.h"
#include "mulle-objc-class-initialize.h"
#include "mulle-objc-class-lookup.h"
#include "mulle-objc-class-search.h"
#include "mulle-objc-class-struct.h"
#include "mulle-objc-csvdump.h"
#include "mulle-objc-fastclasstable.h"
#include "mulle-objc-fastmethodtable.h"
#include "mulle-objc-infraclass.h"
#include "mulle-objc-ivar.h"
#include "mulle-objc-ivarlist.h"
#include "mulle-objc-kvccache.h"
#include "mulle-objc-load.h"
#include "mulle-objc-metaclass.h"
#include "mulle-objc-method.h"
#include "mulle-objc-methodidconstants.h"
#include "mulle-objc-methodlist.h"
#include "mulle-objc-object.h"
#include "mulle-objc-objectheader.h"
#include "mulle-objc-object-convenience.h"
#include "mulle-objc-property.h"
#include "mulle-objc-propertylist.h"
#include "mulle-objc-retain-release.h"
#include "mulle-objc-signature.h"
#include "mulle-objc-super.h"
#include "mulle-objc-symbolizer.h"
#include "mulle-objc-taggedpointer.h"
#include "mulle-objc-try-catch-finally.h"
#include "mulle-objc-universe.h"
#include "mulle-objc-universe-class.h"
#include "mulle-objc-universe-exception.h"
#include "mulle-objc-universe-fail.h"
#include "mulle-objc-universe-global.h"
#include "mulle-objc-universe-struct.h"
#include "mulle-objc-uniqueid.h"
#include "mulle-objc-uniqueidarray.h"
#include "mulle-objc-version.h"
#include "mulle-objc-walktypes.h"



// add some functions to mulle-vararg for ObjC

#define mulle_vararg_count_ids( args, obj) \
   mulle_vararg_count_pointers( args, obj)
#define mulle_vararg_count_objects( args, obj) \
   mulle_vararg_count_pointers( args, obj)

#define mulle_vararg_next_id( args) \
   mulle_vararg_next_pointer( args, id)
#define mulle_vararg_next_object( args, type) \
   mulle_vararg_next_pointer( args, type)

#define mulle_vararg_push_id( args, value) \
   _mulle_vararg_push( args, id, value)
#define mulle_vararg_push_object( args, value) \
   mulle_vararg_push_id( args, value)

//
// add this before return of int main( argc, char **argc) when running
// named universes and leak checking
//
static inline void  mulle_objc_global_finish( void)
{
   struct _mulle_objc_universe   *universe;

   universe = mulle_objc_global_get_universe_inline( __MULLE_OBJC_UNIVERSEID__);
   if( universe && ! _mulle_objc_universe_is_default( universe) && universe->config.pedantic_exit)
      _mulle_objc_universe_release( universe);
}



#ifdef __has_include
# if __has_include( "_mulle-objc-runtime-versioncheck.h")
#  include "_mulle-objc-runtime-versioncheck.h"
# endif
#endif


#include <assert.h>
#include <stdarg.h>
#include <setjmp.h>

#endif

