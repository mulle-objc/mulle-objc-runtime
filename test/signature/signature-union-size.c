//
//  main.c
//  test-runtime-2
//
//  Created by Nat! on 19/11/14.
//  Copyright (c) 2014 Mulle kybernetiK. All rights reserved.
//
#ifndef __MULLE_OBJC__
# define __MULLE_OBJC_NO_TPS__
# define __MULLE_OBJC_FCS__
# if defined( DEBUG) || ! defined( __OPTIMIZE__)
#  define __MULLE_OBJC_TAO__
# else
#  define __MULLE_OBJC_NO_TAO__
# endif
#endif



#include <mulle-objc-runtime/mulle-objc-runtime.h>

#include <stdio.h>


union test_0
{
   char    c;
};

union test_1
{
   char    c;
   short   s;
};

union test_2
{
   char    c;
   int     i;
};

union test_3
{
   char    c;
   double  d;
};

union test_4
{
   char    c;
   char    c2;
};


union test_5
{
   short   s;
   double  d;
};


union test_6
{
   union test_0  x;
   union test_0  y;
};


union test_7
{
   short   s;
   char    c;
};

union test_8
{
   int     i;
   char    c;
};

union test_9
{
   double  d;
   char    c;
};


union test_10
{
   double  d;
   short   s;
};


union test_11
{
   union test_1   x1;
   union test_2   x2;
   union test_7   x7;
   union test_8   x8;
};


static struct test
{
   char          *signature;
   size_t        size;
   unsigned int  align;
} tests[] =
{
   { "(?=c)",  sizeof( union test_0), alignof( union test_0) },
   { "(?=cs)", sizeof( union test_1), alignof( union test_1) },
   { "(?=ci)", sizeof( union test_2), alignof( union test_2) },
   { "(?=cd)", sizeof( union test_3), alignof( union test_3) },
   { "(?=cc)", sizeof( union test_4), alignof( union test_4) },
   { "(?=sd)", sizeof( union test_5), alignof( union test_5) },
   { "(?=(?=c)(?=c))", sizeof( union test_6), alignof( union test_6) },
   { "(?=sc)", sizeof( union test_7), alignof( union test_7) },
   { "(?=ic)", sizeof( union test_8), alignof( union test_8) },
   { "(?=dc)", sizeof( union test_9), alignof( union test_9) },
   { "(?=ds)", sizeof( union test_10), alignof( union test_10) },
   { "(?=(?=cs)(?=ci)(?=sc)(?=ic))", sizeof( union test_11), alignof( union test_11) },
   { 0 }
};



static void   dummy_callback( char *type, struct mulle_objc_typeinfo *info, void *userinfo)
{
}



static void   test( struct test *p)
{
   char                         *next;
   struct mulle_objc_typeinfo   type_info;

   printf( "type=\"%s\" (size=%zd, align=%u): ", p->signature, p->size, p->align);

   next = _mulle_objc_type_parse( p->signature,
                                  0,
                                  &type_info,
                                  _mulle_objc_signature_supply_scalar_typeinfo,
                                  dummy_callback,
                                  NULL);
   if( ! next)
   {
      printf( "SYNTAX ERROR IN SIGNATURE: \"%s\"\n", p->signature);
      return;
   }

   if( type_info.natural_size != p->size)
   {
      printf( "DIFFERENCE IN SIZE: \"%zd\" vs expected \"%zd\"\n",
                  type_info.natural_size, p->size);
      return;
   }
   if( type_info.natural_alignment != p->align)
   {
      printf( "DIFFERENCE IN ALIGN: \"%u\" vs expected \"%u\"\n",
                  type_info.natural_alignment, p->align);
      return;
   }
   printf( "OK\n");
}




int   main( int argc, const char * argv[])
{
   struct test  *p;

   for( p = tests; p->signature; p++)
   {
      test( p);
   }

   return 0;
}


