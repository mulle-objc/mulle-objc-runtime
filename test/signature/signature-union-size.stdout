type="(?=c)" (size=1, align=1): OK
type="(?=cs)" (size=2, align=2): OK
type="(?=ci)" (size=4, align=4): OK
type="(?=cd)" (size=8, align=8): OK
type="(?=cc)" (size=1, align=1): OK
type="(?=sd)" (size=8, align=8): OK
type="(?=(?=c)(?=c))" (size=1, align=1): OK
type="(?=sc)" (size=2, align=2): OK
type="(?=ic)" (size=4, align=4): OK
type="(?=dc)" (size=8, align=8): OK
type="(?=ds)" (size=8, align=8): OK
type="(?=(?=cs)(?=ci)(?=sc)(?=ic))" (size=4, align=4): OK
